### DISCUSSION TITLE
jupyterlab-markup - a plugin for markdown-it in jupyterlab

### DISCUSSION THREAD

**Discussion by choldgraf** [executablebooks/meta#171] [Show and tell]
@bollwyvl just pointed me to [`jupyterlab-markup`](https://github.com/agoose77/jupyterlab-markup) from @agoose77, which seems to add functionality for markdown-it in JupyterLab. I don't know whether it is designed to be extensible for new plugins (e.g. the ones for MyST markdown), but perhaps it's something to look at for getting MyST support in JupyterLab.

**Comment by bollwyvl** [comment]
I did some fiddling around to get to something extensible, about 95% like the current marked output with off-the-shelf plugins: task lists, rewritten `./` links to `/files/`, anchored headers, etc.

- [PR](https://github.com/agoose77/jupyterlab-markup/pull/10) 
- [binder](https://mybinder.org/v2/gh/bollwyvl/jupyterlab-markup/add-binder?urlpath=lab%2Ftree%2Findex.ipynb)

With the way markdown-it loads plugins, making it generally extensible looks like about ten lines of boilerplate per simple plugin (this could be tightened up), while still being _able_ to do complex ones with lots of config options.

My biggest concern, at least for notebooks, is that there would need to be a way to document _which_ extensions are needed to properly render a document in a way that would be portable between renderers (nbconvert, lab, etc). My thinking is this would end up being a JSON schema which could eventually land in nbformat, a la:

```yaml
metadata:
  jupyter:
    commonmark:
      extensions:
        - task-list
        - header-anchors
cells: []
```

So you'd at least get a _warning_ that something wasn't going to be right. As to _where_ that list came from, and how one might get it for their py/js/r/whatever renderer, remains mysterious, as the community/spec doesn't seem to have any sort of way to do this in a document, out of band, or whatever.

**Reply by chrisjsewell** [reply]
Yep cheers, looks promising 👍

**Reply by choldgraf** [reply]
good point about needing a metadata spec to define what extra markdown syntax may be inside. This seems like a much more principled way of doing it than just letting people put whatever random markdown they want in the notebooks :-)

**Reply by bollwyvl** [reply]
raised as an issue about metadata over there... https://github.com/agoose77/jupyterlab-markup/issues/13

Also, the [PR](https://github.com/agoose77/jupyterlab-markup/pull/10) is looking pretty good, and has grown user configurable settings... a spin through the binder and notes would be greatly appreciated!

![Screenshot from 2020-11-03 14-08-47](https://user-images.githubusercontent.com/45380/98029969-4b683d80-1dde-11eb-9894-45d6bc186c9e.png)

**Reply by agoose77** [reply]
@bollwyvl is very kind, but most of the input is his! 🙏
